add interfaces that wrap the libev prepare, check, and idle watchers

garlick opened the pull request:
This PR tentatively implements prepare, check, and idle watchers directly wrapping the libev ones.  See [libev docs](http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod) for how these work.

I've also added a variant of `flux_reactor_start()` called `flux_reactor_run()` which optionally takes the FLUX_REACTOR_NONBLOCK flag.  If this flag is passed, the reactor handles all outstanding events and, if there are still active watchers, returns -1 with errno set to EWOULDBLOCK.

This is in response to issues raised in #362.  I took the shotgun approached and provided all three of the watchers discussed plus the "run once" interface.  I'm not sure if all of this is needed.  Perhaps @trws can play around with this branch and see what really works for his use case and then we can decide what makes the cut, what to do with `flux_reactor_start()` etc.

grondo commented:
My guess is in addition to FLUX_REACTOR_NONBLOCK we want something more like `uv_run`'s RUN_ONCE, which runs at least one handler before returning to the caller. With NONBLOCK (which is like `uv_run (NOWAIT)` if there are no events you return immediately back to user so you can't really call it in a loop or you'd be busywaiting...

garlick commented:
Good point, and libev has a similar flag so I've added that.

garlick commented:
Just pushed some additional commits that
* Use the modern libev names for flags: NOWAIT and ONCE, instead of NONBLOCK and ONESHOT.
* Try implementing a single `flux_watcher_t` type for prepare/check/idle, with the same callback prototype, and the same destroy/start/stop functions.
* Apply this to all other watcher types for fun, except the msg_watchers which are slightly more work (wanted to get feedback before tackling that one)
* Update in-tree users

This attempt at consolidating the watcher API falls short of the general approaches in #363.  Perhaps this is a good incremental step?

Feedback welcome.

trws commented:
Thanks Jim, this looks really good!  I'll work on some testing and 
integration with python/cap soon.


On 15 Sep 2015, at 14:56, Jim Garlick wrote:

> Just pushed some additional commits that
> * Use the modern libev names for flags: NOWAIT and ONCE, instead of 
> NONBLOCK and ONESHOT.
> * Try implementing a single `flux_watcher_t` type for 
> prepare/check/idle, with the same callback prototype, and the same 
> destroy/start/stop functions.
> * Apply this to all other watcher types for fun, except the 
> msg_watchers which are slightly more work (wanted to get feedback 
> before tackling that one)
> * Update in-tree users
>
> This attempt at consolidating the watcher API falls short of the 
> general approaches in #363.  Perhaps this is a good incremental step?
>
> Feedback welcome.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/407#issuecomment-140557970

garlick commented:
More coming - since I was in there, I'm taking the opportunity to untangle message dispatcher and reactor.  The net result should be that one reactor should be able to handle multiple flux_t handles, e.g. supporting async rpc's on either.  Plus the code is much easier to understand with that split out.

garlick commented:
This cleanup worked out fairly well IMHO.

This could use a bit of review if someone has time, especially the API.  If it looks good I'll write some tests for multiple handles, and at long last produce some documentation for the reactor/dispatch API.

grondo commented:
I haven't had time to do a full review, but so far this is great. I really like how it is shaping up.
Having the watcher types consolidated is a huge win, and obviously being able to watch more than one handle is a big step forward.

Some very early comments (sorry if some of these are just plain wrong, I'm only doing a quick review for now)

Name for `flux_reactor_add` seemed a bit odd, since your adding a new handle into the reactor maybe it should be `flux_reactor_add_handle` (not sure about this one, but "add" seemed a bit generic)

From the commit message on  71a5477 it sounds like you have to add the second handle before registering any message watcher? It would seem like in a dynamic application or module you might have handles to other instances coming and going, so you might want to be able to add and remove handle watchers at any time. (So do we need `flux_reactor_remove_handle ()`?)
I'm not entirely sure this is a valid complaint though.

I hate to even say this, but it does kind of seem like having the reactor be embedded in a handle is turning out to be a little backwards? Just throwing it out there, but should we think about having a `flux_reactor_t *` which is a convenience layer around libev and handles generic watchers like fd/timer/idle/prepare/check/signal/etc, and can additionally watch any number of handles and invoke their associated message dispatchers? (I apologize if this is a dumb suggestion) :fearful:

garlick commented:
Thanks for the good comments @grondo!

I actually like the `flux_reactor_t *` idea as the handle/reactor thing does seem like an unholy alliance in light of the multiple handles per reactor thing.  It'll cost some verbosity though...
```
flux_t h = flux_open()
flux_reactor_t *r = flux_reactor_create();
flux_watcher_t *w = flux_msg_watcher_create (h, match, cb, arg);

flux_watcher_start (r, w); // implicitly starts h handle watcher
flux_reactor_run (r)
flux_watcher_stop (r, w) // implicitly stops h handle watcher

flux_watcher_destroy (w);
flux_reactor_destroy (r);
flux_close (h);
```

garlick commented:
The constraint on adding message watchers is that the second handle needs to have its reactor "redirected" before adding _its_ first message watcher, so that the dynamically created dispatcher attaches its internal handle watcher to the correct reactor.  The first handle could already be running.

If we exposed the reactor as discussed above, that constraint would go away of course.

trws commented:
Hmm... For some reason, seeing the reactor laid out that way puts me in mind of queue handlers in [Grand Central Dispatch](https://developer.apple.com/library/ios/documentation/Performance/Reference/GCD_libdispatch_Ref/#//apple_ref/c/func/dispatch_queue_create).  Since most use-cases will result in a user having just one reactor that they care about, but potentially more than one handle, might it make sense to just have a way to request that thread/process's reactor?  Using the C11 syntax, since pthreads TLS is obtuse enough I would have to look it up, it might be something like this.

```c
// reactor.c
thread_local flux_reactor_t * main_reactor = NULL;
flux_reactor_t * flux_get_main_reactor(){
  if (!main_reactor)
    main_reactor = flux_reactor_create();
  return main_reactor
}//cleanup registered for thread exit...

// User code
flux_t h = flux_open(flux_get_main_reactor()); // supply reactor to handle
flux_watcher_t *w = flux_msg_watcher_create (h, match, cb, arg);

flux_watcher_start (w); // implicitly starts h handle watcher on main_reactor
flux_reactor_run (flux_get_main_reactor()); // or flux_reactor_wait, if the reactor is running in another thread perhaps?
flux_watcher_stop (w); // implicitly stops h handle watcher on main_reactor

flux_watcher_destroy (w);
flux_close (h);
```

It could also be flipped around, adding handles to the reactors rather than this way, but it would tend to streamline things.

garlick commented:
I like that idea but I don't think we can assume either C11 or TLS is going to work in all users of the public API...

trws commented:
It's not a big deal, the pass a reactor to a handle model might be an option even so, but I'm curious what you mean @garlick.  We depend on linux pthreads, which implements {get,set}specific and TLS both with the older GCC and newer C11 versions.  If we decide to target OS X, or some of the more interesting BSDs, the pthreads version wouldn't be available, but the c11/GCC versions would.  Worse to worst it would be pretty trivial to implement with a hash table and a lock, or better with a concurrent hash table or tree.

garlick commented:
Oops, sorry, you're right.  I was thinking of applications that might not be linked against pthreads but since `libflux-core.s` is, TLS should be fine.

garlick commented:
Just pushed a change that decouples the reactor and handle, and just presumes there at most one reactor per thread/process, created on demand in TLS.

The main reactor functions were able to drop their `flux_t` argument e.g.
```C
int flux_reactor_run (int flags);
void flux_reactor_stop (void);
void flux_reactor_stop_error (void);

typedef void (*flux_watcher_f)(flux_watcher_t *w, int revents, void *arg);
void flux_watcher_destroy (flux_watcher_t *w);
void flux_watcher_start (flux_watcher_t *w);
void flux_watcher_stop (flux_watcher_t *w);
```

The `flux_reactor_add()` function went away since all dispatchers implicitly register their handle watcher with the one implied reactor for their context.

The commit is larger than it should be because I (stupidly) dropped some of the compat stuff and updated a bunch of users.  That doesn't mean I'll be resistant to backing up and going another direction if people feel I've gone too far down the "implied reactor" path.

garlick commented:
Yeah, having slept on it. I don't think I'm happy with the "implied reactor".  Unless others disagree, I think I'll back up and go with a `flux_reactor_t *` that has to be created/destroyed by the user.  The reactor functions and callbacks would all have a `flux_reactor_t *` argument (replacing what was originally a `flux_t` argument).

`flux_rpc_then()` would add a `flux_reactor_t *` argument.

`kvs_watch()` (or its redesigned successor) would also add one.

Alternatively we could add a function to associate a `flux_t` handle with a reactor, and leave those as is, but I think the additional argument might actually improve clarity.  For example, not everybody immediately groks that `kvs_watch()` callbacks don't get called if the reactor isn't started.

trws commented:
I'm a bit in-between on this.  I like having the option of letting the user create one if they want a separate reactor, but the watcher_* functions aside from create don't really need one, they have a watcher that must have been associated with a handle at some point.  Same thing with the handles, unless we want to support the same watcher being in multiple reactors, or the same handle in multiple reactors, I don't really see the reason for the repetition.  If we do want to allow one thing to be added to multiple reactors, then maybe it would make sense to have a `flux_reactor_add(reactor, watcher)` that takes watchers, and just never directly associate a handle with a watcher at all, since each watcher must have been created with a handle. Something like...

```c
flux_t h = flux_open();
flux_reactor_t *r = flux_get_main_reactor();
flux_watcher_t *w = flux_msg_watcher_create (h, match, cb, arg);

flux_reactor_add(r, w); // explicitly starts watcher w, and thus h handle watcher on main_reactor
flux_reactor_run (r); // or flux_reactor_wait, if the reactor is running in another thread perhaps?
flux_reactor_remove (r, w); // removes w from r, and implicitly stops h handle watcher on main_reactor

flux_watcher_destroy (w);
flux_close (h);
```

grondo commented:
`watcher_t`'s like timers, fd watchers, signal watchers, and anything else we'd like to bring in from libev don't need to be associated with a handle. This could allow use of flux reactor in an application without having called `flux_open()`.

It feels like the straightforward way to think about this is that we have reactors that can watch timers, fds, etc, and/or one or more handles. Message watchers are associated with dispatchers which are part of handles.

So you create a reactor (or use magic `flux_get_main_reactor()`) and add watchers and/or handles. Message "watchers" are added to handles. As a handle gets added/removed from a reactor, the dispatcher is automatically added/removed from the reactor.

Is this too simplistic a view?

trws commented:
It makes sense to me at least.

garlick commented:
Yes that all corresponds with my view, except (and this may not be exactly what you meant), registering a handle watcher might be done without involving a dispatcher.  A handle watcher is like a fd watcher except `flux_send()` and `flux_recv()` are used instead of `write()` and `read()`.

Instead, registering a message watcher with a handle should maybe start/stop an internal handle watcher wired to the dispatcher callback.

I guess then there needs to be another function to associate a handle's dispatcher with the reactor like
```
flux_dispatch_start (flux_reactor_t *r, flux_t h);
```

garlick commented:
Oh, and if the `flux_dispatch_start()` call isn't made, the watcher could be started on `flux_get_main_reactor()` on the first message watcher start.

grondo commented:
I'm getting a bit lost now, sorry!

It kind of seems like the message "dispatcher" is hidden from the user behind the handle. I could be wrong though because I haven't had time to give the new code a good review, sorry! However, if this is the case, I'd prefer not to have dispatcher functions bubble up to the API.

Could you get rid of that call by:
 * On addition of a handle to a reactor (sorry not sure of the function name here), the dispatcher is "started" thereby starting all associated message watchers.
 * On start of a message watcher for a handle already on a reactor this does what you'd expect (message watcher immediately activated)
 * On handle removal from a watcher, dispatcher is "stopped" thereby stopping all associated message watchers

This seems to do what a user would expect, and leaves less room for accidentally adding a handle to the reactor and forgetting to call `flux_dispatch_start`.

However, I may have missed or forgotten about some obvious things.

grondo commented:
Oops, I see you already commented on forgetting `flux_dispatch_start`. Sorry I missed that!

garlick commented:
Right, I tossed out `flux_dispatch_start()` as the way of adding a handle to the reactor that is distinct from a handle watcher.  (Er, it _is_ a handle watcher, but the callback and arg are supplied internally by the dispatcher, hidden as you say).

Rename and then we're on the same page?

garlick commented:
Another idea is to have `flux_open()` accept a FLUX_O_REACTOR flag that creates a reactor for use by the dispatcher, and destroyed when the handle is destroyed.  Message watchers could then be added to the handle as now.  Then provide
```
flux_reactor_t *flux_get_reactor (flux_t h);
void flux_set_reactor (flux_t h, flux_reactor_t *r);
```
for the case where other type of watchers are needed, or the handle needs to join the reactor of another handle.

The flux_t handle that is pre-created for comms modules could be created with FLUX_O_REACTOR so the boilerplate code for a module does not need to change.

I like this a bit better than the `flux_get_main_reactor()` approach.

trws commented:
That's not so far from my initial thought, except that the handle implicitly has a reactor rather than supplying it with one from outside.  Having the handle be the main way to manage all of this seems fine.  One question would be though, is it valid to have a handle that has no reactor? If not, could it just always create one to start, or require one to be passed as an argument to open?

garlick commented:
Yeah, that was inspired by your idea @trws.

A handle with no reactor should be valid, e.g. for a utility that exclusively uses only synchronous RPC, only subscribes to events, etc..

grondo commented:
I also like the proposal above a bit better than `flux_get_main_reactor(void)`  

It is almost exactly what we have now, which already seems to work well, but promotes the reactor to a named member of a handle. Am I right in saying you would "watch" messages from two handles by something like:
```c
 flux_set_reactor (h2, flux_get_reactor (h1));
 flux_run (h1); // or flux_reactor_run (flux_get_reactor (h1)) ?
```

The only slightly confusing thing I see for users is that the design seems a bit strange when you have to pass a handle to the reactor and a reactor to the handle to watch messages.
However, in this case it seems warranted and at this point I like the design better than what we have now.

trws commented:
I'm good with this, with the caveat that I think this code:
```c
flux_t h = flux_open(NULL, 0);
flux_reactor_run(h);
```
Should run a reactor on the handle, even if one hasn't been explicitly created in the handle by flag or set.  Really that's all I meant earlier, and it's a good point that if the user isn't going to run the reactor they shouldn't need to explicitly create one.

garlick commented:
Yes I was thinking
```C
flux_set_reactor (h2, flux_get_reactor (h1));
flux_reactor_run (flux_get_reactor (h1));
```

To watch messages I was thinking
```C
flux_t h = flux_open (NULL, FLUX_O_REACTOR);
flux_reactor_t *r = flux_get_reactor (h);
flux_watcher_t *w = flux_msg_watcher_create (h, match, callback, arg);
flux_watcher_start (r, w);
...
flux_reactor_run (r, 0);
...
flux_watcher_destroy (w);
flux_close (h);
```

I suppose actually it is unncessary to have the FLUX_O_REACTOR flag.  We could just create it on demand at `flux_get_reactor(h)` if one hasn't been set.

This is still quite different than what we had before because the reactor functions and non-message callbacks will all take a `flux_reactor_t *` argument where they had a `flux_t` before.

trws commented:
I like that overall, and this is really a minor quibble, but in this 
model it really looks like the `flux_watcher_start` is actually a method 
acting on the reactor.  Maybe keep the semantics but call it 
`flux_reactor_add` or flip the arguments around?

On 18 Sep 2015, at 13:58, Jim Garlick wrote:

> Yes I was thinking
> ```C
> flux_set_reactor (h2, flux_get_reactor (h1));
> flux_reactor_run (flux_get_reactor (h1));
> ```
>
> To watch messages I was thinking
> ```C
> flux_t h = flux_open (NULL, FLUX_O_REACTOR);
> flux_reactor_t *r = flux_get_reactor (h);
> flux_watcher_t *w = flux_msg_watcher_create (h, match, callback, arg);
> flux_watcher_start (r, w);
> ...
> flux_reactor_run (r, 0);
> ...
> flux_watcher_destroy (w);
> flux_close (h);
> ```
>
> I suppose actually it is unncessary to have the FLUX_O_REACTOR flag.  
> We could just create it on demand at `flux_get_reactor(h)` if one 
> hasn't been set.
>
> This is still quite different than what we had before because the 
> reactor functions and non-message callbacks will all take a 
> `flux_reactor_t *` argument where they had a `flux_t` before.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/407#issuecomment-141567582

garlick commented:
OK, I'll flip the args for `flux_watcher_start/stop()`.  Thanks.

garlick commented:
OK, I backed up and reworked so that `flux_reactor_t` is exposed and is used to start/stop the reactor, and start/top watchers.

I renamed `flux_msg_watcher_t` to `flux_handler_t` (too generic? ) and did not integrate it into the unified watcher type, which underscores the fact that it's a callback of the dispatcher not the reactor.  Start/stop and the callback take a `flux_t` handle not a `flux_reactor_t *`.
```
typedef struct flux_handler flux_handler_t;

typedef void (*flux_handler_f)(flux_t h, flux_handler_t *w,
                               const flux_msg_t *msg, void *arg);

flux_handler_t *flux_handler_create (const struct flux_match match,
                                     flux_handler_f cb, void *arg);

void flux_handler_destroy (flux_handler_t *w);

void flux_handler_start (flux_handler_t *w, flux_t h);
void flux_handler_stop (flux_handler_t *w, flux_t h);
```
Internally, the dispatcher registers a handle watcher on whatever reactor has been associated with the `flux_t` handle, by calling `flux_get_reactor()`, which creates a reactor on demand if one has not already been associated with `flux_set_reactor()`.

Other high level functions like `flux_rpc_then()`, `flux_reduce_create()`, and `kvs_watch()` that internally register reactor watchers or handlers continue to use the handle's reactor, as described above.

When there are multiple `flux_t` handles open and they need to share a reactor, the procedure is as discussed above:
```
flux_set_reactor (h2, flux_get_reactor (h1));
```
after that, handlers registered on either h1 or h2, including those created for RPC's, `kvs_watch()`, etc. will just work via h1's reactor.

Feedback needed, then after any iteration I'll do man pages and perhaps some more tests.

garlick commented:
(Sorry just forced a push after noticing something I missed in my rebase).  All done for now.

grondo commented:
I think this is great! 

I agree that `flux_handler_t` seems a bit too generic at first (especially with the name similarity between flux "handle" and flux "handler" ) However, if the object might later be extended to more than just messages then maybe it is appropriate? If not then either `flux_msg_handler_t` or `flux_msg_consumer_t`? .. not sure I have a good suggestion actually

garlick commented:
Some random thoughts on replacement names for `flux_handler_t`:

Mailbox theme:  `flux_mailbox_t`, `flux_msgslot_t`, `flux_mbox_t`.

Consumer of information theme:  `flux_subscriber_t`, `flux_listener_t`, `flux_reader_t`.

Ok, I'm out of ideas for now.

trws commented:
I like the way it reads, though I admit I'm not sure I completely understand the break between message handlers and the others.  If using a timer watcher say, in a reactor that is associated with a handle, is there a way to get the handle without passing it through `arg`?

garlick commented:
The reactor now doesn't have to be watching any handles, or it may be watching multiple handles.

The dispatcher on the other hand is directly associated with one handle.

garlick commented:
(sorry, to answer your question, no you would need to the handle through arg).

garlick commented:
How about `flux_observer_t`?  as in [observer pattern](https://en.wikipedia.org/wiki/Observer_pattern).  See [stackoverflow discussion](http://stackoverflow.com/questions/27619556/name-for-sender-receiver-of-messages)

trws commented:
For me, that's the best idea yet.

On 22 Sep 2015, at 19:49, Jim Garlick wrote:

> How about `flux_observer_t`?  as in [observer 
> pattern](https://en.wikipedia.org/wiki/Observer_pattern).  See 
> [stackoverflow 
> discussion](http://stackoverflow.com/questions/27619556/name-for-sender-receiver-of-messages)
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/407#issuecomment-142475348

grondo commented:
> How about flux_observer_t? as in observer pattern. See stackoverflow discussion

I saw that too, but at first didn't feel like the model fit since it sounded like observers have to be registered  with the object their observing, which observes changes in the object, not general messages.

Looking further it almost seems like an observable might be a nice higher level concept?
http://docs.couchbase.com/developer/java-2.0/observables.html

I'm definitely not an expert in this area though and would defer to you and certainly @trws.

garlick commented:
Maybe `flux_receiver_t` would have less baggage?

I don't really have a good feeling for whether `flux_observer_t` would send somebody more familiar with that pattern looking for semantics that aren't there in Flux.

grondo commented:
Actually, last night I realized you were probably talking about the "observer" observing the `flux_t` handle, in which case it isn't quite as objectionable to me. 

However, going back a few steps, what was wrong with calling these what they are, e.g. flux_msg_handlers or flux_msg_callbacks? I probably missed a plan to use this type for other things besides messages? (I don't mean to be a stick in the mud, but are we making this more complicated than it needs to be? (more than half my fault proabably))

trws commented:
I think we can reasonably say that `flux_t` and `flux_reactor_t` are "observable" or "publishers," and that what has most recently been called `flux_handler_t` and `flux_watcher_t` are each "observers" or "subscribers" to those interfaces.  This is one of the things that gets me about the interfaces as they are a little bit though.  These are really dispatchers, rather than observables, where each may have several observables served by it, they aren't a single generic slot each.  The current interface creates an observable and adds one observer to it for each watcher, rather than creating an observable and maintaining a list of observers to notify.  In a strict OO setup, the pattern would probably end up looking like this:

```c++
namespace flux {
// Type alias, an observer is any callable object or function pointer with
// signature void (observable *, void*)
class observable;
using observer = std::function<void (observable *, void *)>;
class observable {
    struct ob_pair {
        observer o;
        void *data;
    };
    std::list<ob_pair> observers;
public:
    void register_observer(observer o, void *data)
    {
        observers.push_front(ob_pair{o, data});
    }
    void unregister_observer(observer o, void *data)
    {
        observers.remove(ob_pair{o, data});
    }
protected:
    void notify_all()
    {
        for (auto pair : observers) {
            pair.o(this, pair.data);
        }
    }
};

class timer_observable : public observable {
    ev_loop *loop;
    std::chrono::duration frequency;
    ev_timer timer;
    static void timer_cb (struct ev_loop *loop, ev_timer *tw, int revents)
    {
        timer_observable *to = reinterpret_cast<timer_observable>(tw->data);
        to->notify_all();
    }

public:
    timer_observable(std::chrono::duration<double, std::chrono::microseconds> after,
                     std::chrono::duration<double, std::chrono::microseconds> repeat)
        : frequency(frequency)
    {
        ev_timer_init(&timer, timer_cb, after.count(), repeat.count());
        timer.data = this;
    }
};
}
```

For the record, I really prefer the new names for this pattern based on passing functions or functors rather than objects around, which would be Event and Delegate, so I'll use those here.  Basically, the reactor would be an argument used to create an object that generates notifications on an event, or an Event.  Then delegates would be added to the event.  In this world, the handler or receiver or whatever would be treated the same way as everything else.  The resulting interface would look something like this in C.

```c
typedef struct flux_event flux_event_t;
typedef struct flux_delegate {
    flux_delegate_f * cb;
    void* data;
} flux_delegate_t;
typedef void (*flux_delegate_f)(flux_event_t *r, int revents, void *arg);

// message event, notify on this message type on this handle using this
// reactor
flux_event_t *flux_message_event_create (flux_t h, const flux_msg_t *msg);

// fd watcher
flux_event_t *flux_fd_event_create (flux_reactor_t *r int fd, int events);
int flux_fd_event_get_fd (flux_event_t *w);

flux_event_add_delegate(flux_delegate_t d);
// Remove only one that matches fully
flux_event_remove_delegate(flux_delegate_t d);


void flux_event_destroy (flux_event_t *w);
void flux_event_start (flux_event_t *w);
void flux_event_stop (flux_event_t *w);

//usage

void read_fd(flux_event_t *e, int revents, void *arg){
    printf("got something on fd %d\n", flux_fd_event_get_fd(e));
}
flux_event_t * fe = flux_fd_event_create(r, 0, 3);
flux_event_add_delegate((flux_delegate){read_fd,NULL});
flux_event_start(fe);
```

The upside is everything is consistent, the downside is that it means two calls to register a delegate for a new event.  Another, possible, upside is that it would be easy to use this to give the same notification to multiple interested parties in a given unit of the program, rather than only the most recent registrant.  Anyway, this has more been about me thinking it out than anything, sorry for the huge post. =/

garlick commented:
@trws do you mean a "handle watcher" would be one form of event (formerly watcher I guess) and a "message watcher" would be a delegate of the handle watcher?  So then the dispatcher code for mapping events to delegates would be part of the internal implementation of the handle watcher.  In other words, the reactor notifies the event when a message is ready on the handle, and the event itself reads a message, thereby consuming the event, and dispatches it to delegate(s).

If that's the case then this seems like not a great fit for the other types of watchers where we have no notion of how to consume the event and distribute it to multiple delegates.

I'll give this all another read when I get to work - it might just take a while to sink in.

trws commented:
Message is actually still the odd one a bit.  When you create a timer event, say, all the delegates will be receiving notifications every time the associated timer fires.  It's so trivial as to be able to use the default implementation of the notifier in the OO version.  The handle/message end is a bit more complicated, because the current message watchers receive only a subset of the events fired.  To make everything even, there would probably be a `flux_comm_event` or similar that fires an event on every single event produced by a handle, but that's more or less what the handle itself represents anyway right?  A `flux_message_event` then could be produced by a delegate on the handle event, or as a convenience directly off of something larger.  I conflated these in the example interface earlier because I was thinking we would always want the filtered version, and that the reactor is part of the handle.  Expanding it out could end up something like this:

```c
// Delegates on this get called for *every* message
flux_comm_event_t *flux_comm_event_create (flux_reactor_t r, flux_t h);
// Above implements a filter to provide a subset of messages like this
flux_message_event_t *flux_message_event_create (flux_comm_event_t *e, const flux_msg_t *msg);
```

To make it all easier to reason about, it helps to have the type-safety in the pointers for this part of it, since message events are actually themselves delegates on comm events.  Alternately there could be a second version of add_delegate that allows for filtering, then maybe what's going on is a little more clear?

```c
typedef void flux_event_match_t
// Delegates on this get called for *every* message
flux_event_t *flux_comm_event_create (flux_reactor_t r, flux_t h);
void flux_event_add_delegate_where(flux_event_t *e, const flux_event_match_t* filter);

// usage
flux_event_t *e = flux_comm_event_create (r, h);
flux_event_add_filtered_delegate(e, &FLUX_MSG_ANY);
```

garlick commented:
The "filtered delegate" does make sense, although would this style of interface be more appropriate in a language binding built on top of the base C one?  As a C/systems guy this feels a little off to me, and I'm not sure if it addresses a tangible problem with the original proposal.

Feel free to argue with me on this @trws, but in the interest of getting the milestone work knocked off, what if we simply agreed on a rename for `flux_handler_t` to say `flux_msg_handler_t` and moved forward?

I apologize if it seems like I'm being lazy or provincial; I hope not, so please argue with me if I'm not seeing the clear benefit of retooling for event/delegate/filter_delegate.

trws commented:
Nope, I'm perfectly good with doing a rename and pleading bigger fish to fry.  The only urge to prod at this at all is that so many of our watchers are messages it seems a shame to have them be the odd one out, that can't be treated generically for start/stop/destroy.  As with the rest, if you don't deem it worth it or disagree feel free to just call this done, but I think all that would take would be to hold a reference to the reactor or handle in the watcher and drop the reactor/handle argument from the start/stop functions right?  Unless the idea is to do create then start with multiple different handles/reactors?

garlick commented:
Ah, sorry for not addressing dropping the `flux_reactor_t` argument from start/stop.  You're right, we probably don't need to support multiple reactor loops and could dispense with that argument to the watcher.  So interface would look like this:
```C
typedef struct flux_watcher flux_watcher_t;
typedef void (*flux_watcher_f)(flux_reactor_t *r, flux_watcher_t *w,
                               int revents, void *arg);
void flux_watcher_destroy (flux_watcher_t *w);
void flux_watcher_start (flux_watcher_t *w);
void flux_watcher_stop (flux_watcher_t *w);

// fd watcher
flux_watcher_t *flux_fd_watcher_create (flux_reactor_t *r,
                                        int fd, int events,
                                        flux_watcher_f cb, void *arg);
```

Unfortunately I think we would still need to keep the `flux_reactor_t` arg to `flux_watcher_f` since the callback may need to use it to call `flux_reactor_stop()` or `flux_reactor_stop_error()`.  Alternatively, since that's probably not the common case, we could drop it and provide an accessor, e.g.
```C
flux_reactor_t *flux_watcher_get_reactor (flux_watcher_t *w);
```

The `flux_msg_handler_t` could similarly drop the `flux_t` handle from start/stop and add it to create.

garlick commented:
On message watchers being the odd one out, I think the argument boils down to calling them the odd one out because they really don't fit the reactor watcher model as is, versus replacing the reactor model with something more nuanced that fits everything.  The past practice of pretending they are reactor watchers without changing the model is, I think we all agree, incorrect.  Agreed?

trws commented:
Agreed. They can be treated like them after creation with enough abstraction, but they're really not quite the same thing. I agree that the callback does need access to a reactor by the way, but since it is necessarily part of the watcher and or handle invoking the callback it doesn't really seem like a problem.



Sent with Good (www.good.com)

________________________________
From: Jim Garlick
Sent: Wednesday, September 23, 2015 11:19:09 AM
To: flux-framework/flux-core
Cc: Scogland, Thomas Richard William
Subject: Re: [flux-core] add interfaces that wrap the libev prepare, check, and idle watchers (#407)


On message watchers being the odd one out, I think the argument boils down to calling them the odd one out because they really don't fit the reactor watcher model as is, versus replacing the reactor model with something more nuanced that fits everything. The past practice of pretending they are reactor watchers without changing the model is, I think we all agree, incorrect. Agreed?

—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/407#issuecomment-142686220>.

garlick commented:
OK, just pushed a rework as discussed, and rebased on current master.  Watchers look like this now:

```C
typedef struct flux_watcher flux_watcher_t;
typedef void (*flux_watcher_f)(flux_reactor_t *r, flux_watcher_t *w,
                               int revents, void *arg);
void flux_watcher_destroy (flux_watcher_t *w);
void flux_watcher_start (flux_watcher_t *w);
void flux_watcher_stop (flux_watcher_t *w);

// fd watcher
flux_watcher_t *flux_fd_watcher_create (flux_reactor_t *r, int fd, int events,
                                        flux_watcher_f cb, void *arg);
int flux_fd_watcher_get_fd (flux_watcher_t *w);
```

and the message handlers:
```C
typedef struct flux_msg_handler flux_msg_handler_t;

typedef void (*flux_msg_handler_f)(flux_t h, flux_msg_handler_t *w,
                                   const flux_msg_t *msg, void *arg);

flux_msg_handler_t *flux_msg_handler_create (flux_t h,
                                             const struct flux_match match,
                                             flux_msg_handler_f cb, void *arg);

void flux_msg_handler_destroy (flux_msg_handler_t *w);

void flux_msg_handler_start (flux_msg_handler_t *w);
void flux_msg_handler_stop (flux_msg_handler_t *w);
```

I'll get started on some man pages, though we could certainly iterate some more if necessary!  Feedback welcome.

grondo commented:
Looks good to me!

trws commented:
Same here! :+1:

On 23 Sep 2015, at 16:05, Mark Grondona wrote:

> Looks good to me!
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/407#issuecomment-142756225

garlick commented:
OK, there's an update rebased on current master, and with a full complement of man pages.

I think mergeable at this point.

trws commented:
Looks good to me.  Odd question that popped into my head glancing 
through docs though.  Is a watcher automatically stopped on destroy, or 
must a user stop it first?

On 24 Sep 2015, at 15:28, Jim Garlick wrote:

> OK, there's an update rebased on current master, and with a full 
> complement of man pages.
>
> I think mergeable at this point.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/407#issuecomment-143066491

garlick commented:
The user must stop it first.  Hmm, should I fix that?  Even if not I should probably cover it in the docs.

grondo commented:
Yeah, sounds like it would cause a crash otherwise?

trws commented:
A note in the doc would be fine by me, especially if it's a complicated fix to change the behavior. Making it either error or stop the thing on destroy would be good for resilience to user error though.

garlick commented:
Yeah the reactor would still have a reference if it is destroyed without being stopped.

The downside to having destroy also stop is that _that_ would crash if the reactor was already destroyed.  Probably less likely though, I'll go ahead and fix it that way and update the docs accordingly.

grondo commented:
This is pure laziness on my part, but for utilities and such I thought it would be nice if watchers had an option to be destroyed when the reactor is destroyed. That can be a later enhancement though, if it is even a good idea at all...

trws commented:
Apropos of @grondo's comment I'll just leave this here... #139 [talloc](https://talloc.samba.org/talloc/doc/html/index.html)

grondo commented:
@trws: *touché* ;-)

garlick commented:
Perhaps for another day?   I have reactor fatigue (sounds fatal doesn't it?)

grondo commented:
Yeah, definitely for another day. Reactor fatigue is a serious issue covered by NRC guide 06-11 Revision 1.

trws commented:
Yup.   All looking good to me. :+1:

garlick commented:
+1 for accurate NRC reference!

grondo commented:
Hm, about to merge, but what is a good summary of all the work done here for the merge commit?

garlick commented:
Is "improve reactor/dispatch interfaces and refactor internals" good enough?

grondo commented:
Yes.

garlick commented:
Forced update coming.  I'll squash those last commits, plus I missed a silly backslash in the Makefile.am which caused the clean target to work much like teenagers cleaning the bathroom.

grondo commented:
Hm, something strange in [travis logs]( https://travis-ci.org/flux-framework/flux-core/jobs/82062844) that we might want to check out:
```
not ok 31 - coproc segfaulted
#   Failed test 'coproc segfaulted'
#   at ../../../../src/common/libutil/test/coproc.c line 202.
# Looks like you planned 30 tests but ran 31.
# Looks like you failed 1 test of 31 run.
PASS: test_coproc.t
```

I haven't looked at the test, maybe that is expected? But looks like the test program should have failed but didn't?

garlick commented:
Uh oh, OK, I'll look into that one!

garlick commented:
I opened #415 on multiple problems with this test.

I didn't have time to go back through travis this afternoon but I suspect it may be a racy test that (due to a secondary problem with the test) doesn't stop make check.  I'll kick off another travis build and see what happens.

I did touch this code, but only to update it for RFC 7, and didn't see anything wrong as I reviewed it again just now.

grondo commented:
Looks like the issue #415 happens only under clang perhaps.
Anyway, this may have been happening even before this PR, so I'm going to go ahead an merge.
